kimchi/src/circuits/polynomials/xor.rs

```diff
@@ -0,0 +1,546 @@
+//~ |      0 | copy     `in1`   | copy     `in1'`  |
+//~ |      1 | copy     `in2`   | copy     `in2'`  |
+//~ |      2 | copy     `out`   | copy     `out'`  |
+//~ |      3 | plookup0 `in1_0` |                  |
+//~ |      4 | plookup1 `in1_1` |                  |
+//~ |      5 | plookup2 `in1_2` |                  |
+//~ |      6 | plookup3 `in1_3` |                  |
+//~ |      7 | plookup0 `in2_0` |                  |
+//~ |      8 | plookup1 `in2_1` |                  |
+//~ |      9 | plookup2 `in2_2` |                  |
+//~ |     10 | plookup3 `in2_3` |                  |
+//~ |     11 | plookup0 `out_0` |                  |
+//~ |     12 | plookup1 `out_1` |                  |
+//~ |     13 | plookup2 `out_2` |                  |
+//~ |     14 | plookup3 `out_3` |                  |
+//~
+//~ One single gate with next values of `in1'`, `in2'` and `out'` being zero can be used to check
+//~ that the original `in1`, `in2` and `out` had 16-bits. We can chain this gate 4 times as follows
+//~ to obtain a gadget for 64-bit words XOR:
+//~
+//~  | Row | `CircuitGate` | Purpose                                    |
+//~  | --- | ------------- | ------------------------------------------ |
+//~  |   0 | `Xor16`       | Xor 2 least significant bytes of the words |
+//~  |   1 | `Xor16`       | Xor next 2 bytes of the words              |
+//~  |   2 | `Xor16`       | Xor next 2 bytes of the words              |
+//~  |   3 | `Xor16`       | Xor 2 most significant bytes of the words  |
+//~  |   4 | `Zero`        | Zero values, can be reused as generic gate |
+//~
+//~ ```admonition::notice
+//~  We could half the number of rows of the 64-bit XOR gadget by having lookups
+//~  for 8 bits at a time, but for now we will use the 4-bit XOR table that we have.
+//~  Rough computations show that if we run 8 or more Keccaks in one circuit we should
+//~  use the 8-bit XOR table.
+//~ ```
+#[derive(Default)]
+pub struct Xor16<F>(PhantomData<F>);
+
+impl<F> Argument<F> for Xor16<F>
+where
+    F: PrimeField,
+{
+    const ARGUMENT_TYPE: ArgumentType = ArgumentType::Gate(GateType::Xor16);
+    const CONSTRAINTS: u32 = 3;
+
+    // Constraints for Xor16
+    //   * Operates on Curr and Next rows
+    //   * Constrain the decomposition of `in1`, `in2` and `out` of multiples of 16 bits
+    //   * The actual XOR is performed thanks to the plookups of 4-bit XORs.
+    fn constraint_checks<T: ExprOps<F>>(env: &ArgumentEnv<F, T>) -> Vec<T> {
+        // Returns the constraints:
+        // in1 = in1_0 + in1_1 * 2^4 + in1_2 * 2^8 + in1_3 * 2^12 + next_in1 * 2^16
+        // in2 = in2_0 + in2_1 * 2^4 + in2_2 * 2^8 + in2_3 * 2^12 + next_in2 * 2^16
+        // out = out_0 + out_1 * 2^4 + out_2 * 2^8 + out_3 * 2^12 + next_out * 2^16
+        (0..3)
+            .map(|i| {
+                env.witness_curr(i)
+                    - chunk_sum(env, 3 + 4 * i)
+                    - T::from(2u64).pow(16) * env.witness_next(i)
```

Comment:
I'm not sure that it helps to have `chunk_sum` here, it might be clearer to write out the constraints explicitly like they are above.

Reply:
Oh well, I was trying to avoid code duplication, but i guess it is not that bad since it's just 3 times. I will change it prior merge.